File: python/printing.py
import os

from documents import FlowchartDiagram


class PhysicalPrinter:
    def __init__(self):
        self._is_available = False
        self._job_count = 0

    @property
    def is_available(self):
        return self._is_available

    @property
    def job_count(self):
        return self._job_count

    def start_document(self, is_summary, is_pdf, name):
        raise NotImplementedError("Can't call this from a unit test")

    def end_document(self):
        raise NotImplementedError("Can't call this from a unit test")

class PrintQueue:
    def __init__(self, physical_printer : PhysicalPrinter):
        self._printer = physical_printer
        self._items = []

    def add(self, data):
        self._items.append(data)

class PrintMetadata:
    def __init__(self, info_file_type : str):
        self._info_file_type = info_file_type
        self.filename = self._generate_temp_filename(info_file_type)

    def _generate_temp_filename(self, info_file_type) -> str:
        temp_filename = os.path.join(os.path.dirname(os.path.abspath(__file__)), "tempfile")
        return f"{temp_filename}.{info_file_type}"

class DiagramPrintDriver:
    def __init__(self):
        self._diagram = None

    def set_diagram(self, diagram : FlowchartDiagram):
        self._diagram = diagram

    def release_diagram(self):
        self._diagram = None

    def print_to(self, physical_printer : PhysicalPrinter) -> bool:
        raise NotImplementedError("Can't call this from a unit test")

class PrinterDriverFactory:
    _instance = None

    @staticmethod
    def get_instance():
        if PrinterDriverFactory._instance is None:
            PrinterDriverFactory._instance = PrinterDriverFactory()
        return PrinterDriverFactory._instance

    # For unit tests
    @staticmethod
    def set_instance(instance):
        PrinterDriverFactory._instance = instance

    def create_driver_for_print(self) -> DiagramPrintDriver:
        return DiagramPrintDriver()

File: typescript/src/printing.ts
import { FlowchartDiagram } from './documents';
import * as path from 'path';
import * as fs from 'fs';

enum Toner {
  Black,
  Cyan,
  Magenta,
  Yellow
}

class PhysicalPrinter {
  private _isAvailable: boolean = false;
  private _jobCount: number = 0;
  private _tonerLevels: Map<Toner, number>;

  constructor() {
    this._tonerLevels = new Map([
      [Toner.Black, 0],
      [Toner.Cyan, 0],
      [Toner.Magenta, 0],
      [Toner.Yellow, 0]
    ]);
  }

  get tonerLevels(): Map<Toner, number> {
    return this._tonerLevels;
  }

  get isAvailable(): boolean {
    return this._isAvailable;
  }

  get jobCount(): number {
    return this._jobCount;
  }

  startDocument(isSummary: boolean, isPdf: boolean, name: string): boolean {
    throw new Error("Can't call this from a unit test");
  }

  endDocument(): void {
    throw new Error("Can't call this from a unit test");
  }
}

class PrintQueue {
  private _printer: PhysicalPrinter;
  private _items: any[] = [];

  constructor(physicalPrinter: PhysicalPrinter) {
    this._printer = physicalPrinter;
  }

  add(data: any): void {
    this._items.push(data);
  }
}

class PrintMetadata {
  private _infoFileType: string;
  filename: string;

  constructor(infoFileType: string) {
    this._infoFileType = infoFileType;
    this.filename = this._generateTempFilename(infoFileType);
  }

  private _generateTempFilename(infoFileType: string): string {
    const tempFilename = path.join(path.dirname(__filename), "tempfile");
    return `${tempFilename}.${infoFileType}`;
  }
}

class DiagramPrintDriver {
  private _diagram: FlowchartDiagram | null = null;

  setDiagram(diagram: FlowchartDiagram): void {
    this._diagram = diagram;
  }

  releaseDiagram(): void {
    this._diagram = null;
  }

  printTo(physicalPrinter: PhysicalPrinter): boolean {
    throw new Error("Can't call this from a unit test");
  }
}

class PrinterDriverFactory {
  private static _instance: PrinterDriverFactory | null = null;

  static getInstance(): PrinterDriverFactory {
    if (PrinterDriverFactory._instance === null) {
      PrinterDriverFactory._instance = new PrinterDriverFactory();
    }
    return PrinterDriverFactory._instance;
  }

  // For unit tests
  static setInstance(instance: PrinterDriverFactory): void {
    PrinterDriverFactory._instance = instance;
  }

  createDriverForPrint(): DiagramPrintDriver {
    return new DiagramPrintDriver();
  }
}

export {
  Toner,
  PhysicalPrinter,
  PrintQueue,
  PrintMetadata,
  DiagramPrintDriver,
  PrinterDriverFactory
};
